class-two-factor-core.php

```diff
@@ -857,6 +898,69 @@ public static function verify_login_nonce( $user_id, $nonce ) {
 		return false;
 	}
 
+	/**
+	 * Determine the minimum wait between two factor attempts for a user.
+	 *
+	 * This implements an increasing backoff, requiring an attacker to wait longer
+	 * each time to attempt to brute-force the login.
+	 *
+	 * @param WP_User $user The user being operated upon.
+	 * @return int Time delay in seconds between login attempts.
+	 */
+	public static function get_user_time_delay( $user ) {
+		/**
+		 * Filter the minimum time duration between two factor attempts.
+		 *
+		 * @param int     $rate_limit The number of seconds between two factor attempts.
+		 */
+		$rate_limit  = apply_filters( 'two_factor_rate_limit', 5 );
+
+		$user_failed_logins = get_user_meta( $user->ID, self::USER_FAILED_LOGIN_ATTEMPTS_KEY, true );
+		if ( $user_failed_logins ) {
+			$rate_limit = pow( 2, $user_failed_logins ) * $rate_limit;
+
+			// Limit to 1 hour maximum time delay.
+			$rate_limit = min( HOUR_IN_SECONDS, $rate_limit );
```

Comment:
It's probably worth adding a filter here too, do you have any notion of what you'd expect for the max timeout?

I'm leaning towards 15 minutes, as that's ~10 failures at 1s incremental rate limit, 1hr is only 12 failures, so maxing out at 15 mins seems reasonable to me.

Reply:
A filter sounds like a good idea, and 15 minutes sounds reasonable 👍🏻 We can iterate before the `0.8.0` release based on feedback from w.org users, if needed.

Reply:
We may eventually want to add something where the counter resets to `USER_FAILED_LOGIN_ATTEMPTS_KEY` is cleared out after ~2 hours of inactivity. Otherwise a legit user who's having lots of problems logging in could end up in a situation where they have to wait 15 minutes between every single attempt, which would be pretty frustrating.

I don't think that's a blocker for merging this, though.